### ISSUE TITLE
Should fips and fips-updates be mutually exclusive?

### ISSUE THREAD

**Issue by OddBloke** [CanonicalLtd/ubuntu-advantage-client#636]
After enabling fips-updates, `ua status` shows:

```
Account: UA Client Test
Subscription: UA Client Test Contract
Valid until: n/a
Technical support level: advanced

SERVICE
cc-eal        entitled         inactive
cis-audit     none             n/a
esm           entitled         n/a
fips          entitled         inactive
fips-updates  entitled         active
livepatch     entitled         n/a

Enable entitlements with `ua enable <service>`
```

Should we allow people to enable `fips` if `fips-updates` is enabled? (Given the definition of FIPS, I worry that people may not realise the impact that will have on their compliance.)

**Comment by powersj** [comment]
If you activate fips-update, you should not be able to active fips.
If you activate fips, then (and I'd like feedback here) you should be able to activate fips-updates?

Can you confirm with Joy?

**Comment by joylatten** [comment]
Use of fips and fips-updates in the tool should be mutually exclusive. If fips is enabled, then implicitly fips-updates PPA is disabled or removed. If fips-updates in enabled, then implicitly fips PPA is disabled or removed.

It is ok to go from fips to fips-update within the tool. However, fips-update to fips implies a downgrade since fips-updates packages usually supercede those in fips. So within the tool, when someone goes from fips-updates to fips, it is not allowed and point them to a URL with manual instructions of how to downgrade fips packages. Security Certs team will document how to downgrade FIPS packages. The team will also document how to manually disable fips. These things will not be done in the tool.
 
If there are disable flags in the tool for fips and fips-updates, they can generate a message  pointing to this URL on how to disable or downgrade fips packages.

**Comment by OddBloke** [comment]
@joylatten Thanks for commenting! I have one immediate question (though I'm sure more will follow once I've thought this through a bit more).

> Use of fips and fips-updates in the tool should be mutually exclusive. If fips is enabled, then implicitly fips-updates PPA is disabled or removed. If fips-updates in enabled, then implicitly fips PPA is disabled or removed.

In `ubuntu-advantage-client` parlance, "disabling" and "removing" are synonymous (i.e. `ua disable esm` deconfigures the configured ESM repo from your system). When you say "disable" here, do you mean it in that sense, or do you mean "prevent it from being enabled"?

**Comment by cmars** [comment]
I would find it kind of weird to have fips-updates, but no fips. Or why would I want a fips without updates? I'm probably missing a lot of background and context and the definition of what these entitlements actually are, but I think it is pretty confusing.

**Comment by urosj** [comment]
Looks like we’re starting to put business logic into otherwise quite “dumb” client up to now, and that seems a bit concerning if this is not isolated. Probably quite innocent in the case of fips as this exclusion rule is highly unlikely to change.

**Comment by powersj** [comment]
FIPS and FIPS updates are two different things:

- FIPS is the fully certificated set of software. Some customers want the certified set to say they are running what is certified and do not wish to deviate away from this. There are issues with auditing if they do.
- For customers who wish to apply security fixes between certifications Canonical updates the linux kernel every 3 weeks and the user space modules are updated quarterly. These modules are built on FIPS approved packages, but are not certified. This is FIPS updates.

As @joylatten stated, going from FIPS to FIPS updates is possible, but the user needs to understand that they are no longer certified. For some customers that is an issue. However, going from FIPS updates to FIPS is not really possible since you would be downgrading software.

I am uncertain how anything but the client would present this to the user, but I am open to ideas.

**Comment by OddBloke** [comment]
@urosj At the moment, I think we're just trying to determine what the correct behaviour is. (We could feasibly have this business logic in the server if we defined mutually exclusive entitlements generically!)

**Comment by cmars** [comment]
Maybe "fips-certified" and "fips-compatibility-updates" would be more appropriate?

**Comment by joylatten** [comment]
In all honesty it is ok to first have fips and then include fips-updates to get fips with security bugfixes. It is most likely that fips-updates will always supercede fips. so works out ok. But as @powersj first pointed out, if you activate fips-updates, is not a good idea to activate fips.

**Comment by joylatten** [comment]
yeah sorry about introducing additional logic into it... I was thinking to avoid a downgrade because apt wont do anything since it will say already at newest version. What does the new tool say or do in this case?  Maybe its just enough to report this back to the user.
